File: Data Structures & Algorithms/two-integer-sum/submission-3.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        // value, index
        unordered_map<int, int> seen; 
        for (int i = 0; i < nums.size(); i++) {
            int d = target - nums[i];

            if (seen.contains(d)) {
                return {seen[d], i};
            }

            seen[nums[i]] = i;
        }
        
    }
};

File: Data Structures & Algorithms/search-2d-matrix/submission-3.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # m * n matrix
        m = len(matrix)
        n = len(matrix[0])

        l = 0
        r = m * n - 2

        while l <= r:
            mid = l + (r - l) // 2
            x = mid // m 
            y = mid & m 

            n = matrix[x][y]

            if n == target:
                return True
            elif n < target:
                l = mid + 1
            else:
                r = mid - 1
        
        return False